File: app/src/main/java/com/example/mg/masterdetail/WeatherApplication.java
package com.example.mg.masterdetail;

import android.app.Activity;
import android.app.Application;
import android.content.Context;

import com.example.mg.masterdetail.DI.DaggerIAppComponent;
import com.example.mg.masterdetail.DI.IAppComponent;
import com.example.mg.masterdetail.DI.Modules.AppModule;
import com.example.mg.masterdetail.DI.Modules.NetworkModule;
import com.squareup.leakcanary.LeakCanary;
import com.squareup.leakcanary.RefWatcher;

public class WeatherApplication extends Application {
    private RefWatcher mRefWatcher;
    private IAppComponent appComponent;

    public static WeatherApplication get(Activity activity) {
        return (WeatherApplication) activity.getApplication();
    }

    public static RefWatcher getRefWatcher(Context context) {
        WeatherApplication application = (WeatherApplication) context.getApplicationContext();
        return application.mRefWatcher;
    }

    public IAppComponent getAppComponent() {
        return appComponent;
    }


    @Override
    public void onCreate() {
        super.onCreate();
        initLeakCanary();
        initDagger();
    }

    private void initDagger() {
        appComponent = DaggerIAppComponent.builder()
                .appModule(new AppModule(this))
                .networkModule(new NetworkModule())
                .build();
        appComponent.inject(this);
    }

    private void initLeakCanary() {
        if (LeakCanary.isInAnalyzerProcess(this)) {
            // This process is dedicated to LeakCanary for heap analysis.
            // You should not init your app in this process.
            return;
        }
        mRefWatcher = LeakCanary.install(this);
    }


}

File: app/src/main/java/com/example/mg/masterdetail/WeatherApplication.kt
package com.example.mg.masterdetail

import android.app.Activity
import android.app.Application
import android.content.Context
import com.example.mg.masterdetail.di.appModule
import com.example.mg.masterdetail.di.dayViewModelModule
import com.example.mg.masterdetail.di.networkModule
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher
import org.koin.android.ext.koin.androidContext
import org.koin.core.context.startKoin

class WeatherApplication : Application() {
    private var mRefWatcher: RefWatcher? = null

    companion object {
        operator fun get(activity: Activity): WeatherApplication {
            return activity.application as WeatherApplication
        }

        fun getRefWatcher(context: Context): RefWatcher? {
            val application = context.applicationContext as WeatherApplication
            return application.mRefWatcher
        }
    }

    override fun onCreate() {
        super.onCreate()
        initLeakCanary()
        initKoin()
    }

    private fun initLeakCanary() {
        if (LeakCanary.isInAnalyzerProcess(this)) {
            // This process is dedicated to LeakCanary for heap analysis.
            // You should not init your app in this process.
            return
        }
        mRefWatcher = LeakCanary.install(this)
    }

    private fun initKoin() {
        startKoin {
            androidContext(this@WeatherApplication)
            modules(listOf(appModule, networkModule, dayViewModelModule))
        }
    }
}